servers/physics_2d/body_pair_2d_sw.cpp

```diff
@@ -288,21 +288,17 @@ bool BodyPair2DSW::setup(real_t p_step) {
 		if (A->is_shape_set_as_one_way_collision(shape_A)) {
 			Vector2 direction = xform_A.get_axis(1).normalized();
 			bool valid = false;
-			if (B->get_linear_velocity().dot(direction) >= 0) {
-				for (int i = 0; i < contact_count; i++) {
-					Contact &c = contacts[i];
-					if (!c.reused) {
-						continue;
-					}
-					if (c.normal.dot(direction) > 0) { //greater (normal inverted)
-						continue;
-					}
-
-					valid = true;
-					break;
+			for (int i = 0; i < contact_count; i++) {
+				Contact &c = contacts[i];
+				if (!c.reused) {
+					continue;
 				}
+				if (c.normal.dot(direction) >= 0) { //greater (normal inverted)
```

Comment:
In order to get more consistent results with side collisions, the best would be to add a little threshold to this check so it's not affected by float precision around 0.

Something like:
```
#define ONE_WAY_DIR_THRESHOLD 0.001

// A case
if (c.normal.dot(direction) > -ONE_WAY_DIR_THRESHOLD) //greater (normal inverted)

// B case
if (dot < ONE_WAY_DIR_THRESHOLD) //less (normal ok)
```

Compared to the current PR's version, I'm getting similar results, and it allows rotations of 180 degrees to pass as well.

Reply:
This part of the code was not changed by this PR. Only the indenting from removing the `if` statement (line 291) above:

![Screenshot_2021-01-06_17-06-34](https://user-images.githubusercontent.com/9253928/103798548-c7327080-5041-11eb-8f18-f185af3b5cbd.png)

My [testing](https://github.com/madmiraal/godot-gym) has also shown that this is not required. Do you have an MRP where this is required?

Bear in mind that this is an edge case, and, in general, `RigidBodies` are continuously moving and rotating, so most users wouldn't notice the difference. Furthermore, as [discussed](https://github.com/godotengine/godot/pull/42574#issuecomment-735372133) above, addressing the correct behaviour for rotated bodies is more involved than just ensuring that 90° rotated bodies (which precision errors don't actually allow us to do) behave consistently.

Ultimately this should be consistent with `KinematicBody`; so unless we decide that addressing the rotated bodies issues needs to be done in this PR, this should be kept as it is.

Reply:
The logic behind my proposed change is to make things consistent when the body hits the side of a platform. From the previous discussion, it was decided that it would always go through in this case. That could be ensured for more cases by adding a little flexibility for mathematical approximation in this check.

The case I'm talking about is not 90°, but 180° rotation (platform upside-down).

You can reproduce it in your test suite (Around the clock, tests 385 & 387).

Or in the new test I've added recently in the physics test framework, by setting a platform angle of 180°:
https://github.com/godotengine/godot-demo-projects/pull/573
This one is mostly based on your Around the Clock test, but you can set parameters more easily to reproduce certain cases.

My suggestion is a little change, it adds to consistency and doesn't seem to break anything. I think the same should apply to KinematicBody movements, but I'm just starting on that case so I can't point out the code to change yet (if any).

Reply:
Also, you *have* changed this code from `> 0` to `>= 0` in this PR, probably for a similar reason (making side collision go through in some other cases) so it makes sense to go a bit further in that direction and ensure more cases to work this way.